### PR TITLE
Introduce stream::select_with_strategy

### PR DESCRIPTION
**Pull request by najamelan**
The current stream::select function only supports round robin. To change that
would be a breaking change, so for now I propose a `select_bias` which
takes a closure that defines the selection strategy.

- When a breaking change is planned, we could re-consider this, but it
  does have 4 generic parameters of which one is a closure, compared to
  the 2 streams in `select`. We could also reimplement `select` on top of this,
  we just need to deal with hiding the generics of the closure, but I reckon a 
  function pointer can be named and we don't need to capture anything....
- note the name select_bias. It's not consistent with `select_biased` used
  elsewhere in the lib. This can obviously be changed, but I feel the "ed"
  at the end does not add any semantic value but makes function names
  longer.
- There isn't really integration tests included right now, but the examples
  in the docs get tested by doc test.

### CODE REVIEW

File: futures-util/src/stream/select.rs [hunk 1]
```diff
@@ -1,50 +1,67 @@
...
 /// streams have completed.
 ///
 /// Note that this function consumes both streams and returns a wrapped
 /// version of them.
+///
+/// ## Examples
+///
+/// ```rust
+/// # futures::executor::block_on(async {
+/// use futures_util::stream::{ repeat, select, StreamExt };
```
Comment (taiki-e) [inline futures-util/src/stream/select.rs:32]: I would prefer to always refer `futures` in examples, not futures-util. (like existing examples)

File: futures-util/src/stream/select.rs [hunk 2]
```diff
@@ -1,50 +1,67 @@
...
-use crate::stream::{Fuse, StreamExt};
+use crate::stream::{select_with_strategy, PollNext, SelectWithStrategy};
 use core::pin::Pin;
 use futures_core::stream::{FusedStream, Stream};
 use futures_core::task::{Context, Poll};
 use pin_project_lite::pin_project;
 
 pin_project! {
     /// Stream for the [`select()`] function.
-    #[derive(Debug)]
```
Comment (taiki-e) [inline futures-util/src/stream/select.rs:10]: Why Debug impl is removed?

Comment (najamelan) [reply futures-util/src/stream/select.rs:10]: ah, sorry. My bad. I intended to put it back but forgot. The reason is this:

```
error: implementation of `Debug` is not general enough
  --> futures-util/src/stream/select.rs:10:14
   |
10 |     #[derive(Debug)]
   |              ^^^^^ implementation of `Debug` is not general enough
   |
   = note: `Debug` would have to be implemented for the type `for<'r> fn(&'r mut PollNext) -> PollNext`
   = note: ...but `Debug` is actually implemented for the type `fn(&'0 mut PollNext) -> PollNext`, for some specific lifetime `'0`
```

I think we need a manual implementation, but I'm not sure what to put in it. I suppose the derive only implemented it when the generics implemented it. We can access inner through SelectWithStrategy to call their Debug I suppose.

Comment (najamelan) [reply futures-util/src/stream/select.rs:10]: We can do this:
```rs
impl<St1, St2> fmt::Debug for Select<St1, St2>
where
    St1: fmt::Debug,
    St2: fmt::Debug,
{
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!( f, "Select over {:?}", self.inner.get_ref() )
    }
}
```

With the two repeat streams of the example it gives:
```
Select over (Repeat { item: 1 }, Repeat { item: 2 })
```
I should probably also add a Debug to `SelectWithStrategy`. What do you think of the above implementation?

Comment (taiki-e) [reply futures-util/src/stream/select.rs:10]: Ah, I think we need to change the `Debug` impl of `SelectWithStrategy` not to print closure (like https://github.com/rust-lang/futures-rs/pull/1677 did). After that, we could re-add `#[derive(Debug)]` to `Select`.

Comment (najamelan) [reply futures-util/src/stream/select.rs]: Ok, makes sense. The result for `Select` is:

`Select { inner: SelectWithStrategy { stream1: Fuse { stream: Repeat { item: 1 }, done: false }, stream2: Fuse { stream: Repeat { item: 2 }, done: false } } }`

File: futures-util/src/stream/select_with_strategy.rs [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+impl<St1, St2, Clos, State> fmt::Debug for SelectWithStrategy<St1, St2, Clos, State>
+where
+    St1: fmt::Debug,
+    St2: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("SelectWithStrategy")
+            .field("stream1", &self.stream1)
+            .field("stream2", &self.stream2)
+            .finish()
```
Comment (taiki-e) [inline futures-util/src/stream/select_with_strategy.rs:226]: It probably makes sense to print the `state` field.

Comment (najamelan) [reply futures-util/src/stream/select_with_strategy.rs]: I have added the state field. It does means implementing `Debug` depends on `State` implementing it.

**Review by taiki-e — APPROVED** [review]
LGTM. Thanks @najamelan!

File: futures-util/src/stream/select_with_strategy.rs [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+pin_project! {
+    /// Stream for the [`select_with_strategy()`] function. See function docs for details.
+    #[must_use = "streams do nothing unless polled"]
+    pub struct SelectWithStrategy<St1, St2, Clos, State> {
+        #[pin]
+        stream1: Fuse<St1>,
+        #[pin]
+        stream2: Fuse<St2>,
+        state: State,
+        clos: Clos,
```
Comment (pickfire) [inline futures-util/src/stream/select_with_strategy.rs:46]: Is this a typo for close?

Comment (najamelan) [reply futures-util/src/stream/select_with_strategy.rs:46]: It's the generic representing the closure. So not a typo. Of course another name can be chosen. I don't really care about what it's called.

**Comment by taiki-e** [comment]
Published in 0.3.16.

**Pull request merged**